src/metrics/index.js

```diff
@@ -0,0 +1,244 @@
+const { METRICS: defaultOptions } = require('../constants')
+const Stats = require('./stats')
+
+const initialCounters = [
+  'dataReceived',
+  'dataSent'
+]
+
+const directionToEvent = {
+  in: 'dataReceived',
+  out: 'dataSent'
+}
+
+class Metrics {
+  /**
+   *
+   * @param {object} options
+   * @param {number} options.computeThrottleMaxQueueSize
+   * @param {number} options.computeThrottleTimeout
+   * @param {Array<number>} options.movingAverageIntervals
+   * @param {number} options.maxOldPeersRetention
+   */
+  constructor (options) {
+    this._options = mergeOptions(defaultOptions, options)
+    this._globalStats = new Stats(initialCounters, this._options)
+    this._peerStats = new Map()
+    this._protocolStats = new Map()
+    this._oldPeers = oldPeerLRU(this._options.maxOldPeersRetention)
+    this._running = false
+    this.onMessage = this.onMessage.bind(this)
+  }
+
+  /**
+   * Must be called for stats to saved. Any data pushed for tracking
+   * will be ignored.
+   */
+  start () {
+    this._running = true
+  }
+
+  /**
+   * Stops all averages timers and prevents new data from being tracked.
+   * Once `stop` is called, `start` must be called to resume stats tracking.
+   */
+  stop () {
+    this._running = false
+    this._globalStats.stop()
+    for (const stats of this._peerStats.values()) {
+      stats.stop()
+    }
+    for (const stats of this._protocolStats.values()) {
+      stats.stop()
+    }
+  }
+
+  /**
+   * Gets the global `Stats` object
+   * @returns {Stats}
+   */
+  get global () {
+    return this._globalStats
+  }
+
+  /**
+   * Returns a list of `PeerId` strings currently being tracked
+   * @returns {Array<string>}
+   */
+  get peers () {
+    return Array.from(this._peerStats.keys())
+  }
+
+  /**
+   * Returns a list of `PeerId` strings currently in the
+   * LRU, disconnected peer cache.
+   * @returns {Array<string>}
+   */
+  get oldPeers () {
+    return Array.from(this._oldPeers.keys())
+  }
+
+  /**
+   * Returns the `Stats` object for the given `PeerId` whether it
+   * is a live peer, or in the disconnected peer LRU cache.
+   * @param {PeerId} peerId
+   * @returns {Stats}
+   */
+  forPeer (peerId) {
+    const idString = peerId.toString()
+    return this._peerStats.get(idString) || this._oldPeers.get(idString)
+  }
+
+  /**
+   * Returns a list of all protocol strings currently being tracked.
+   * @returns {Array<string>}
+   */
+  get protocols () {
+    return Array.from(this._protocolStats.keys())
+  }
+
+  /**
+   * Returns the `Stats` object for the given `protocol`.
+   * @param {string} protocol
+   * @returns {Stats}
+   */
+  forProtocol (protocol) {
+    return this._protocolStats.get(protocol)
+  }
+
+  /**
+   * Should be called when all connections to a given peer
+   * have closed. The `Stats` collection for the peer will
+   * be stopped and moved to an LRU for temporary retention.
+   * @param {PeerId} peerId
+   */
+  onPeerDisconnected (peerId) {
```

Comment:
This should be called

Reply:
Yes, good catch. I will think through some more test cases to catch these things. I'm not doing any "old peer" verification yet.